File: src/main/java/com/github/novicezk/midjourney/controller/SubmitController.java
package com.github.novicezk.midjourney.controller;

import cn.hutool.core.text.CharSequenceUtil;
import cn.hutool.core.util.RandomUtil;
import com.github.novicezk.midjourney.Constants;
import com.github.novicezk.midjourney.ProxyProperties;
import com.github.novicezk.midjourney.ReturnCode;
import com.github.novicezk.midjourney.dto.BaseSubmitDTO;
import com.github.novicezk.midjourney.dto.SubmitBlendDTO;
import com.github.novicezk.midjourney.dto.SubmitChangeDTO;
import com.github.novicezk.midjourney.dto.SubmitDescribeDTO;
import com.github.novicezk.midjourney.dto.SubmitImagineDTO;
import com.github.novicezk.midjourney.dto.SubmitSimpleChangeDTO;
import com.github.novicezk.midjourney.enums.TaskAction;
import com.github.novicezk.midjourney.enums.TaskStatus;
import com.github.novicezk.midjourney.enums.TranslateWay;
import com.github.novicezk.midjourney.exception.BannedPromptException;
import com.github.novicezk.midjourney.result.SubmitResultVO;
import com.github.novicezk.midjourney.service.TaskService;
import com.github.novicezk.midjourney.service.TaskStoreService;
import com.github.novicezk.midjourney.service.TranslateService;
import com.github.novicezk.midjourney.support.Task;
import com.github.novicezk.midjourney.util.BannedPromptUtils;
import com.github.novicezk.midjourney.util.ConvertUtils;
import com.github.novicezk.midjourney.util.MimeTypeUtils;
import com.github.novicezk.midjourney.util.SnowFlake;
import com.github.novicezk.midjourney.util.TaskChangeParams;
import eu.maxschuster.dataurl.DataUrl;
import eu.maxschuster.dataurl.DataUrlSerializer;
import eu.maxschuster.dataurl.IDataUrlSerializer;
import lombok.RequiredArgsConstructor;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RestController;

import java.net.MalformedURLException;
import java.util.ArrayList;
import java.util.List;
import java.util.Optional;
import java.util.Set;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

@RestController
@RequestMapping("/submit")
@RequiredArgsConstructor
public class SubmitController {
	private final TranslateService translateService;
	private final TaskStoreService taskStoreService;
	private final ProxyProperties properties;
	private final TaskService taskService;

	@PostMapping("/imagine")
	public SubmitResultVO imagine(@RequestBody SubmitImagineDTO imagineDTO) {
		String prompt = imagineDTO.getPrompt();
		if (CharSequenceUtil.isBlank(prompt)) {
			return SubmitResultVO.fail(ReturnCode.VALIDATION_ERROR, "prompt cannot be empty");
		}
		prompt = prompt.trim();
		Task task = newTask(imagineDTO);
		task.setAction(TaskAction.IMAGINE);
		task.setPrompt(prompt);
		String promptEn = translatePrompt(prompt);
		try {
			BannedPromptUtils.checkBanned(promptEn);
		} catch (BannedPromptException e) {
			return SubmitResultVO.fail(ReturnCode.BANNED_PROMPT, "may contain sensitive words")
					.setProperty("promptEn", promptEn).setProperty("bannedWord", e.getMessage());
		}
		List<String> base64Array = Optional.ofNullable(imagineDTO.getBase64Array()).orElse(new ArrayList<>());
		if (CharSequenceUtil.isNotBlank(imagineDTO.getBase64())) {
			base64Array.add(imagineDTO.getBase64());
		}
		List<DataUrl> dataUrls;
		try {
			dataUrls = ConvertUtils.convertBase64Array(base64Array);
		} catch (MalformedURLException e) {
			return SubmitResultVO.fail(ReturnCode.VALIDATION_ERROR, "base64 format error");
		}
		task.setPromptEn(promptEn);
		task.setDescription("/imagine " + prompt);
		return this.taskService.submitImagine(task, dataUrls);
	}

	@PostMapping("/simple-change")
	public SubmitResultVO simpleChange(@RequestBody SubmitSimpleChangeDTO simpleChangeDTO) {
		TaskChangeParams changeParams = ConvertUtils.convertChangeParams(simpleChangeDTO.getContent());
		if (changeParams == null) {
			return SubmitResultVO.fail(ReturnCode.VALIDATION_ERROR, "content parameter error");
		}
		SubmitChangeDTO changeDTO = new SubmitChangeDTO();
		changeDTO.setAction(changeParams.getAction());
		changeDTO.setTaskId(changeParams.getId());
		changeDTO.setIndex(changeParams.getIndex());
		changeDTO.setState(simpleChangeDTO.getState());
		changeDTO.setNotifyHook(simpleChangeDTO.getNotifyHook());
		return change(changeDTO);
	}

	@PostMapping("/change")
	public SubmitResultVO change(@RequestBody SubmitChangeDTO changeDTO) {
		if (CharSequenceUtil.isBlank(changeDTO.getTaskId())) {
			return SubmitResultVO.fail(ReturnCode.VALIDATION_ERROR, "taskId cannot be empty");
		}
		if (!Set.of(TaskAction.UPSCALE, TaskAction.VARIATION, TaskAction.REROLL).contains(changeDTO.getAction())) {
			return SubmitResultVO.fail(ReturnCode.VALIDATION_ERROR, "action parameter error");
		}
		String description = "/up " + changeDTO.getTaskId();
		if (TaskAction.REROLL.equals(changeDTO.getAction())) {
			description += " R";
		} else {
			description += " " + changeDTO.getAction().name().charAt(0) + changeDTO.getIndex();
		}
		Task targetTask = this.taskStoreService.get(changeDTO.getTaskId());
		if (targetTask == null) {
			return SubmitResultVO.fail(ReturnCode.NOT_FOUND, "related task does not exist or has expired");
		}
		if (!TaskStatus.SUCCESS.equals(targetTask.getStatus())) {
			return SubmitResultVO.fail(ReturnCode.VALIDATION_ERROR, "related task status error");
		}
		if (!Set.of(TaskAction.IMAGINE, TaskAction.VARIATION, TaskAction.REROLL, TaskAction.BLEND).contains(targetTask.getAction())) {
			return SubmitResultVO.fail(ReturnCode.VALIDATION_ERROR, "related task does not allow variation");
		}
		Task task = newTask(changeDTO);
		task.setAction(changeDTO.getAction());
		task.setPrompt(targetTask.getPrompt());
		task.setPromptEn(targetTask.getPromptEn());
		task.setProperty(Constants.TASK_PROPERTY_FINAL_PROMPT, targetTask.getProperty(Constants.TASK_PROPERTY_FINAL_PROMPT));
		task.setProperty(Constants.TASK_PROPERTY_PROGRESS_MESSAGE_ID, targetTask.getProperty(Constants.TASK_PROPERTY_MESSAGE_ID));
		task.setProperty(Constants.TASK_PROPERTY_DISCORD_INSTANCE_ID, targetTask.getProperty(Constants.TASK_PROPERTY_DISCORD_INSTANCE_ID));
		task.setDescription(description);
		int messageFlags = targetTask.getPropertyGeneric(Constants.TASK_PROPERTY_FLAGS);
		String messageId = targetTask.getPropertyGeneric(Constants.TASK_PROPERTY_MESSAGE_ID);
		String messageHash = targetTask.getPropertyGeneric(Constants.TASK_PROPERTY_MESSAGE_HASH);
		task.setProperty(Constants.TASK_PROPERTY_REFERENCED_MESSAGE_ID, messageId);
		if (TaskAction.UPSCALE.equals(changeDTO.getAction())) {
			return this.taskService.submitUpscale(task, messageId, messageHash, changeDTO.getIndex(), messageFlags);
		} else if (TaskAction.VARIATION.equals(changeDTO.getAction())) {
			return this.taskService.submitVariation(task, messageId, messageHash, changeDTO.getIndex(), messageFlags);
		} else {
			return this.taskService.submitReroll(task, messageId, messageHash, messageFlags);
		}
	}

	@PostMapping("/describe")
	public SubmitResultVO describe(@RequestBody SubmitDescribeDTO describeDTO) {
		if (CharSequenceUtil.isBlank(describeDTO.getBase64())) {
			return SubmitResultVO.fail(ReturnCode.VALIDATION_ERROR, "base64 cannot be empty");
		}
		IDataUrlSerializer serializer = new DataUrlSerializer();
		DataUrl dataUrl;
		try {
			dataUrl = serializer.unserialize(describeDTO.getBase64());
		} catch (MalformedURLException e) {
			return SubmitResultVO.fail(ReturnCode.VALIDATION_ERROR, "base64 format error");
		}
		Task task = newTask(describeDTO);
		task.setAction(TaskAction.DESCRIBE);
		String taskFileName = task.getId() + "." + MimeTypeUtils.guessFileSuffix(dataUrl.getMimeType());
		task.setDescription("/describe " + taskFileName);
		return this.taskService.submitDescribe(task, dataUrl);
	}

	@PostMapping("/blend")
	public SubmitResultVO blend(@RequestBody SubmitBlendDTO blendDTO) {
		List<String> base64Array = blendDTO.getBase64Array();
		if (base64Array == null || base64Array.size() < 2 || base64Array.size() > 5) {
			return SubmitResultVO.fail(ReturnCode.VALIDATION_ERROR, "base64List parameter error");
		}
		if (blendDTO.getDimensions() == null) {
			return SubmitResultVO.fail(ReturnCode.VALIDATION_ERROR, "dimensions parameter error");
		}
		IDataUrlSerializer serializer = new DataUrlSerializer();
		List<DataUrl> dataUrlList = new ArrayList<>();
		try {
			for (String base64 : base64Array) {
				DataUrl dataUrl = serializer.unserialize(base64);
				dataUrlList.add(dataUrl);
			}
		} catch (MalformedURLException e) {
			return SubmitResultVO.fail(ReturnCode.VALIDATION_ERROR, "base64 format error");
		}
		Task task = newTask(blendDTO);
		task.setAction(TaskAction.BLEND);
		task.setDescription("/blend " + task.getId() + " " + dataUrlList.size());
		return this.taskService.submitBlend(task, dataUrlList, blendDTO.getDimensions());
	}

	private Task newTask(BaseSubmitDTO base) {
		Task task = new Task();
		task.setId(System.currentTimeMillis() + RandomUtil.randomNumbers(3));
		task.setSubmitTime(System.currentTimeMillis());
		task.setState(base.getState());
		String notifyHook = CharSequenceUtil.isBlank(base.getNotifyHook()) ? this.properties.getNotifyHook() : base.getNotifyHook();
		task.setProperty(Constants.TASK_PROPERTY_NOTIFY_HOOK, notifyHook);
		task.setProperty(Constants.TASK_PROPERTY_NONCE, SnowFlake.INSTANCE.nextId());
		return task;
	}

	private String translatePrompt(String prompt) {
		if (TranslateWay.NULL.equals(this.properties.getTranslateWay()) || CharSequenceUtil.isBlank(prompt) || !this.translateService.containsChinese(prompt)) {
			return prompt;
		}
		String paramStr = "";
		Matcher paramMatcher = Pattern.compile("\\x20+--[a-z]+.*$", Pattern.CASE_INSENSITIVE).matcher(prompt);
		if (paramMatcher.find()) {
			paramStr = paramMatcher.group(0);
		}
		String promptWithoutParam = CharSequenceUtil.sub(prompt, 0, prompt.length() - paramStr.length());
		List<String> imageUrls = new ArrayList<>();
		Matcher imageMatcher = Pattern.compile("https?://[a-z0-9-_:@&?=+,.!/~*'%$]+\\x20+", Pattern.CASE_INSENSITIVE).matcher(promptWithoutParam);
		while (imageMatcher.find()) {
			imageUrls.add(imageMatcher.group(0));
		}
		String text = promptWithoutParam;
		for (String imageUrl : imageUrls) {
			text = CharSequenceUtil.replaceFirst(text, imageUrl, "");
		}
		if (CharSequenceUtil.isNotBlank(text)) {
			text = this.translateService.translateToEnglish(text).trim();
		}
		if (CharSequenceUtil.isNotBlank(paramStr)) {
			Matcher paramNomatcher = Pattern.compile("--no\\s+(.*?)(?=--|$)").matcher(paramStr);
			if (paramNomatcher.find()) {
				String paramNoStr = paramNomatcher.group(1).trim();
				String paramNoStrEn = this.translateService.translateToEnglish(paramNoStr).trim();
				paramStr = paramNomatcher.replaceFirst("--no " + paramNoStrEn + " ");
			}
		}
		return CharSequenceUtil.join("", imageUrls) + text + paramStr;
	}

}

File: src/controllers/submitController.ts
import { FastifyRequest, FastifyReply } from 'fastify';
import { SubmitResultVO } from '../models/SubmitResultVO';
import { Task } from '../models/Task';
import { TaskAction } from '../enums/TaskAction';
import { TaskStatus } from '../enums/TaskStatus';
import { ReturnCode } from '../constants';
import { SubmitImagineDTO } from '../dto/SubmitImagineDTO';
import { SubmitChangeDTO } from '../dto/SubmitChangeDTO';
import { SubmitSimpleChangeDTO } from '../dto/SubmitSimpleChangeDTO';
import { SubmitDescribeDTO } from '../dto/SubmitDescribeDTO';
import { SubmitShortenDTO } from '../dto/SubmitShortenDTO';
import { SubmitBlendDTO } from '../dto/SubmitBlendDTO';
import { SubmitActionDTO } from '../dto/SubmitActionDTO';
import { SubmitModalDTO } from '../dto/SubmitModalDTO';
import { SubmitEditsDTO } from '../dto/SubmitEditsDTO';
import { TaskService } from '../services/taskService';
import { TaskStoreService } from '../services/store/taskStoreService';
import { TranslateService } from '../services/translate/translateService';
import { config } from '../config';
import { BannedPromptException } from '../exceptions/BannedPromptException';
import { checkBanned } from '../utils/bannedPromptUtils';
import { convertBase64Array, convertChangeParams, getPrimaryPrompt, DataUrl } from '../utils/convertUtils';
import { guessFileSuffix } from '../utils/mimeTypeUtils';
import { SnowFlake } from '../utils/snowflake';
import { TaskChangeParams } from '../utils/taskChangeParams';
import { parseActionFromCustomId } from '../utils/actionUtils';
import { ButtonInfo } from '../utils/buttonUtils';
import {
  TASK_PROPERTY_NOTIFY_HOOK,
  TASK_PROPERTY_NONCE,
  TASK_PROPERTY_FINAL_PROMPT,
  TASK_PROPERTY_PROGRESS_MESSAGE_ID,
  TASK_PROPERTY_DISCORD_INSTANCE_ID,
  TASK_PROPERTY_FLAGS,
  TASK_PROPERTY_MESSAGE_ID,
  TASK_PROPERTY_MESSAGE_HASH,
  TASK_PROPERTY_REFERENCED_MESSAGE_ID,
  TASK_PROPERTY_BUTTONS,
} from '../constants';

/**
 * Submit controller
 */
export class SubmitController {
  private taskService: TaskService;
  private taskStoreService: TaskStoreService;
  private translateService: TranslateService;

  constructor(taskService: TaskService, taskStoreService: TaskStoreService, translateService: TranslateService) {
    this.taskService = taskService;
    this.taskStoreService = taskStoreService;
    this.translateService = translateService;
  }

  async imagine(request: FastifyRequest<{ Body: SubmitImagineDTO }>, reply: FastifyReply): Promise<SubmitResultVO> {
    const imagineDTO = request.body;
    let prompt = imagineDTO.prompt;

    if (!prompt || prompt.trim().length === 0) {
      return SubmitResultVO.fail(ReturnCode.VALIDATION_ERROR, 'prompt cannot be empty');
    }

    prompt = prompt.trim();
    const task = this.newTask(imagineDTO);
    task.action = TaskAction.IMAGINE;
    task.prompt = prompt;

    const promptEn = await this.translatePrompt(prompt);
    try {
      checkBanned(promptEn);
    } catch (e) {
      if (e instanceof BannedPromptException) {
        return SubmitResultVO.fail(ReturnCode.BANNED_PROMPT, 'may contain sensitive words')
          .setProperty('promptEn', promptEn)
          .setProperty('bannedWord', e.message);
      }
      throw e;
    }

    const base64Array = imagineDTO.base64Array || [];
    if (imagineDTO.base64) {
      base64Array.push(imagineDTO.base64);
    }

    let dataUrls;
    try {
      dataUrls = convertBase64Array(base64Array);
    } catch (e) {
      return SubmitResultVO.fail(ReturnCode.VALIDATION_ERROR, 'base64 format error');
    }

    task.promptEn = promptEn;
    task.description = `/imagine ${prompt}`;
    return this.taskService.submitImagine(task, dataUrls);
  }

  async simpleChange(request: FastifyRequest<{ Body: SubmitSimpleChangeDTO }>, reply: FastifyReply): Promise<SubmitResultVO> {
    const simpleChangeDTO = request.body;
    const changeParams = convertChangeParams(simpleChangeDTO.content || '');

    if (!changeParams) {
      return SubmitResultVO.fail(ReturnCode.VALIDATION_ERROR, 'content parameter error');
    }

    const changeDTO: SubmitChangeDTO = {
      action: changeParams.action,
      taskId: changeParams.id,
      index: changeParams.index,
      state: simpleChangeDTO.state,
      notifyHook: simpleChangeDTO.notifyHook,
    };

    // Create a new request-like object with the changeDTO as body
    const changeRequest = {
      ...request,
      body: changeDTO,
    } as FastifyRequest<{ Body: SubmitChangeDTO }>;

    return this.change(changeRequest, reply);
  }

  async change(request: FastifyRequest<{ Body: SubmitChangeDTO }>, reply: FastifyReply): Promise<SubmitResultVO> {
    const changeDTO = request.body;

    if (!changeDTO.taskId) {
      return SubmitResultVO.fail(ReturnCode.VALIDATION_ERROR, 'taskId cannot be empty');
    }

    if (![TaskAction.UPSCALE, TaskAction.VARIATION, TaskAction.REROLL].includes(changeDTO.action!)) {
      return SubmitResultVO.fail(ReturnCode.VALIDATION_ERROR, 'action parameter error');
    }

    let description = `/up ${changeDTO.taskId}`;
    if (changeDTO.action === TaskAction.REROLL) {
      description += ' R';
    } else {
      description += ` ${changeDTO.action!.charAt(0)}${changeDTO.index}`;
    }

    const targetTask = await this.taskStoreService.get(changeDTO.taskId!);
    if (!targetTask) {
      return SubmitResultVO.fail(ReturnCode.NOT_FOUND, 'related task does not exist or has expired');
    }

    if (targetTask.status !== TaskStatus.SUCCESS) {
      return SubmitResultVO.fail(ReturnCode.VALIDATION_ERROR, 'related task status error');
    }

    if (![TaskAction.IMAGINE, TaskAction.VARIATION, TaskAction.REROLL, TaskAction.BLEND].includes(targetTask.action!)) {
      return SubmitResultVO.fail(ReturnCode.VALIDATION_ERROR, 'related task does not allow variation');
    }

    const task = this.newTask(changeDTO);
    task.action = changeDTO.action;
    task.prompt = targetTask.prompt;
    task.promptEn = targetTask.promptEn;
    task.setProperty(TASK_PROPERTY_FINAL_PROMPT, targetTask.getProperty(TASK_PROPERTY_FINAL_PROMPT));
    task.setProperty(TASK_PROPERTY_PROGRESS_MESSAGE_ID, targetTask.getProperty(TASK_PROPERTY_MESSAGE_ID));
    task.setProperty(TASK_PROPERTY_DISCORD_INSTANCE_ID, targetTask.getProperty(TASK_PROPERTY_DISCORD_INSTANCE_ID));
    task.description = description;

    const messageFlags = targetTask.getProperty(TASK_PROPERTY_FLAGS) || 0;
    const messageId = targetTask.getProperty(TASK_PROPERTY_MESSAGE_ID);
    const messageHash = targetTask.getProperty(TASK_PROPERTY_MESSAGE_HASH);
    task.setProperty(TASK_PROPERTY_REFERENCED_MESSAGE_ID, messageId);

    if (changeDTO.action === TaskAction.UPSCALE) {
      return this.taskService.submitUpscale(task, messageId, messageHash, changeDTO.index!, messageFlags);
    } else if (changeDTO.action === TaskAction.VARIATION) {
      return this.taskService.submitVariation(task, messageId, messageHash, changeDTO.index!, messageFlags);
    } else {
      return this.taskService.submitReroll(task, messageId, messageHash, messageFlags);
    }
  }

  async describe(request: FastifyRequest<{ Body: SubmitDescribeDTO }>, reply: FastifyReply): Promise<SubmitResultVO> {
    const describeDTO = request.body;

    if (!describeDTO.base64) {
      return SubmitResultVO.fail(ReturnCode.VALIDATION_ERROR, 'base64 cannot be empty');
    }

    let dataUrl;
    try {
      const dataUrls = convertBase64Array([describeDTO.base64]);
      dataUrl = dataUrls[0];
    } catch (e) {
      return SubmitResultVO.fail(ReturnCode.VALIDATION_ERROR, 'base64 format error');
    }

    const task = this.newTask(describeDTO);
    task.action = TaskAction.DESCRIBE;
    const taskFileName = `${task.id}.${guessFileSuffix(dataUrl.mimeType) || 'png'}`;
    task.description = `/describe ${taskFileName}`;
    return this.taskService.submitDescribe(task, dataUrl);
  }

  async shorten(request: FastifyRequest<{ Body: SubmitShortenDTO }>, reply: FastifyReply): Promise<SubmitResultVO> {
    const shortenDTO = request.body;
    let prompt = shortenDTO.prompt;

    if (!prompt || prompt.trim().length === 0) {
      return SubmitResultVO.fail(ReturnCode.VALIDATION_ERROR, 'prompt cannot be empty');
    }

    prompt = prompt.trim();
    const task = this.newTask(shortenDTO);
    task.action = TaskAction.SHORTEN;
    task.prompt = prompt;

    const promptEn = await this.translatePrompt(prompt);
    task.promptEn = promptEn;
    task.description = `/shorten ${prompt}`;
    return this.taskService.submitShorten(task);
  }

  async blend(request: FastifyRequest<{ Body: SubmitBlendDTO }>, reply: FastifyReply): Promise<SubmitResultVO> {
    const blendDTO = request.body;
    const base64Array = blendDTO.base64Array;

    if (!base64Array || base64Array.length < 2 || base64Array.length > 5) {
      return SubmitResultVO.fail(ReturnCode.VALIDATION_ERROR, 'base64List parameter error');
    }

    if (!blendDTO.dimensions) {
      return SubmitResultVO.fail(ReturnCode.VALIDATION_ERROR, 'dimensions parameter error');
    }

    let dataUrlList;
    try {
      dataUrlList = convertBase64Array(base64Array);
    } catch (e) {
      return SubmitResultVO.fail(ReturnCode.VALIDATION_ERROR, 'base64 format error');
    }

    const task = this.newTask(blendDTO);
    task.action = TaskAction.BLEND;
    task.description = `/blend ${task.id} ${dataUrlList.length}`;
    return this.taskService.submitBlend(task, dataUrlList, blendDTO.dimensions);
  }

  async action(request: FastifyRequest<{ Body: SubmitActionDTO }>, reply: FastifyReply): Promise<SubmitResultVO> {
    const actionDTO = request.body;

    if (!actionDTO.taskId) {
      return SubmitResultVO.fail(ReturnCode.VALIDATION_ERROR, 'taskId cannot be empty');
    }
    if (!actionDTO.customId || actionDTO.customId.trim().length === 0) {
      return SubmitResultVO.fail(ReturnCode.VALIDATION_ERROR, 'customId cannot be empty');
    }

    const targetTask = await this.taskStoreService.get(actionDTO.taskId);
    if (!targetTask) {
      return SubmitResultVO.fail(ReturnCode.NOT_FOUND, 'related task does not exist or has expired');
    }

    if (targetTask.status !== TaskStatus.SUCCESS) {
      return SubmitResultVO.fail(ReturnCode.VALIDATION_ERROR, 'related task status error');
    }

    const task = this.newTask(actionDTO);
    // Parse customId to determine the correct action type
    task.action = parseActionFromCustomId(actionDTO.customId);
    task.prompt = targetTask.prompt;
    task.promptEn = targetTask.promptEn;
    task.setProperty(TASK_PROPERTY_FINAL_PROMPT, targetTask.getProperty(TASK_PROPERTY_FINAL_PROMPT));
    task.setProperty(TASK_PROPERTY_PROGRESS_MESSAGE_ID, targetTask.getProperty(TASK_PROPERTY_MESSAGE_ID));
    task.setProperty(TASK_PROPERTY_DISCORD_INSTANCE_ID, targetTask.getProperty(TASK_PROPERTY_DISCORD_INSTANCE_ID));
    task.description = `/action ${actionDTO.customId}`;

    const messageFlags = targetTask.getProperty(TASK_PROPERTY_FLAGS) || 0;
    const messageId = targetTask.getProperty(TASK_PROPERTY_MESSAGE_ID);
    if (!messageId) {
      return SubmitResultVO.fail(ReturnCode.VALIDATION_ERROR, 'related message not found');
    }

    return this.taskService.submitCustomAction(task, messageId, messageFlags, actionDTO.customId.trim());
  }

  async modal(request: FastifyRequest<{ Body: SubmitModalDTO }>, reply: FastifyReply): Promise<SubmitResultVO> {
    const modalDTO = request.body;

    if (!modalDTO.taskId) {
      return SubmitResultVO.fail(ReturnCode.VALIDATION_ERROR, 'taskId cannot be empty');
    }

    const targetTask = await this.taskStoreService.get(modalDTO.taskId);
    if (!targetTask) {
      return SubmitResultVO.fail(ReturnCode.NOT_FOUND, 'related task does not exist or has expired');
    }

    const task = this.newTask(modalDTO);
    task.action = TaskAction.VARIATION;
    task.prompt = targetTask.prompt;
    task.promptEn = targetTask.promptEn;
    task.setProperty(TASK_PROPERTY_FINAL_PROMPT, targetTask.getProperty(TASK_PROPERTY_FINAL_PROMPT));
    task.setProperty(TASK_PROPERTY_PROGRESS_MESSAGE_ID, targetTask.getProperty(TASK_PROPERTY_MESSAGE_ID));
    task.setProperty(TASK_PROPERTY_DISCORD_INSTANCE_ID, targetTask.getProperty(TASK_PROPERTY_DISCORD_INSTANCE_ID));
    task.description = `/modal ${modalDTO.taskId}`;

    return this.taskService.submitModal(task, {
      prompt: modalDTO.prompt,
      maskBase64: modalDTO.maskBase64,
      modalTaskId: modalDTO.taskId,
    });
  }

  async edits(request: FastifyRequest<{ Body: SubmitEditsDTO }>, reply: FastifyReply): Promise<SubmitResultVO> {
    const editsDTO = request.body;

    if (!editsDTO.taskId) {
      return SubmitResultVO.fail(ReturnCode.VALIDATION_ERROR, 'taskId cannot be empty');
    }

    if (!editsDTO.prompt || editsDTO.prompt.trim().length === 0) {
      return SubmitResultVO.fail(ReturnCode.VALIDATION_ERROR, 'prompt cannot be empty');
    }

    if (!editsDTO.maskBase64 || editsDTO.maskBase64.trim().length === 0) {
      return SubmitResultVO.fail(ReturnCode.VALIDATION_ERROR, 'maskBase64 cannot be empty');
    }

    // Get the target task
    const targetTask = await this.taskStoreService.get(editsDTO.taskId);
    if (!targetTask) {
      return SubmitResultVO.fail(ReturnCode.NOT_FOUND, 'related task does not exist or has expired');
    }

    if (targetTask.status !== TaskStatus.SUCCESS) {
      return SubmitResultVO.fail(ReturnCode.VALIDATION_ERROR, 'related task must be in SUCCESS status');
    }

    // Get message_id from the target task
    const messageId = targetTask.getProperty(TASK_PROPERTY_MESSAGE_ID);
    if (!messageId) {
      return SubmitResultVO.fail(ReturnCode.VALIDATION_ERROR, 'related task does not have a message_id');
    }

    // Extract buttons from the target task
    const buttons = targetTask.getProperty(TASK_PROPERTY_BUTTONS) as ButtonInfo[] | undefined;
    if (!buttons || !Array.isArray(buttons) || buttons.length === 0) {
      return SubmitResultVO.fail(ReturnCode.VALIDATION_ERROR, 'related task does not have button information. The message may not have Inpaint button available.');
    }

    // Find the Inpaint button
    const inpaintButton = buttons.find(btn => 
      btn.customId && btn.customId.toLowerCase().includes('inpaint')
    );

    if (!inpaintButton || !inpaintButton.customId) {
      return SubmitResultVO.fail(ReturnCode.VALIDATION_ERROR, 'Inpaint button not found in the related task');
    }

    let prompt = editsDTO.prompt.trim();
    const task = this.newTask(editsDTO);
    task.action = TaskAction.VARIATION;
    task.prompt = prompt;

    const promptEn = await this.translatePrompt(prompt);
    try {
      checkBanned(promptEn);
    } catch (e) {
      if (e instanceof BannedPromptException) {
        return SubmitResultVO.fail(ReturnCode.BANNED_PROMPT, 'may contain sensitive words')
          .setProperty('promptEn', promptEn)
          .setProperty('bannedWord', e.message);
      }
      throw e;
    }

    task.promptEn = promptEn;
    task.description = `/edits ${prompt}`;
    
    // Set properties from target task
    task.setProperty(TASK_PROPERTY_DISCORD_INSTANCE_ID, targetTask.getProperty(TASK_PROPERTY_DISCORD_INSTANCE_ID));
    task.setProperty(TASK_PROPERTY_PROGRESS_MESSAGE_ID, messageId);
    task.setProperty(TASK_PROPERTY_REFERENCED_MESSAGE_ID, messageId);

    return this.taskService.submitEdits(task, messageId, inpaintButton.customId, editsDTO.maskBase64, promptEn);
  }

  private newTask(base: any): Task {
    const task = new Task();
    task.id = `${Date.now()}${Math.floor(Math.random() * 1000).toString().padStart(3, '0')}`;
    task.submitTime = Date.now();
    task.state = base.state;
    const notifyHook = base.notifyHook || config.mj.notifyHook;
    task.setProperty(TASK_PROPERTY_NOTIFY_HOOK, notifyHook);
    task.setProperty(TASK_PROPERTY_NONCE, SnowFlake.INSTANCE.nextId());
    return task;
  }

  private async translatePrompt(prompt: string): Promise<string> {
    if (config.mj.translateWay === 'NULL' || !prompt || !this.translateService.containsChinese(prompt)) {
      return prompt;
    }

    // Extract parameters
    const paramMatch = prompt.match(/\s+--[a-z]+.*$/i);
    const paramStr = paramMatch ? paramMatch[0] : '';

    const promptWithoutParam = prompt.substring(0, prompt.length - paramStr.length);

    // Extract image URLs
    const imageUrlRegex = /https?:\/\/[a-z0-9-_:@&?=+,.!/~*'%$]+\s+/gi;
    const imageUrls: string[] = [];
    let match;
    while ((match = imageUrlRegex.exec(promptWithoutParam)) !== null) {
      imageUrls.push(match[0]);
    }

    let text = promptWithoutParam;
    for (const imageUrl of imageUrls) {
      text = text.replace(imageUrl, '');
    }

    if (text.trim()) {
      const translated = this.translateService.translateToEnglish(text);
      text = typeof translated === 'string' ? translated.trim() : (await translated).trim();
    }

    // Translate --no parameters
    let translatedParamStr = paramStr;
    if (paramStr) {
      const noMatch = paramStr.match(/--no\s+(.*?)(?=--|$)/);
      if (noMatch) {
        const paramNoStr = noMatch[1].trim();
        const translatedNo = this.translateService.translateToEnglish(paramNoStr);
        const translatedNoStr = typeof translatedNo === 'string' ? translatedNo.trim() : (await translatedNo).trim();
        translatedParamStr = paramStr.replace(noMatch[0], `--no ${translatedNoStr} `);
      }
    }

    return imageUrls.join('') + text + translatedParamStr;
  }
}

